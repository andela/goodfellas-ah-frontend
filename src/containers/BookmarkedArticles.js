import React, { Component } from 'react';
import { connect } from 'react-redux';
import { userPlaceholderImage } from '../mixin';
import { Loader as Loading } from '../components/shared/Loading';
import { fetchProfile } from '../actions/profileActions';
import { getBookmarkedArticles } from '../actions/articleActions';
import BookmarkedArticleList from '../components/article/BookmarkedArticleList';

// eslint-disable-next-line react/prefer-stateless-function
class BookmarkedArticles extends Component {
  componentWillMount = () => {
    const {
      user: { userId },
      fetchProfile: fetchUserProfile,
      getBookmarkedArticles: fetchBookmarkedArticles,
    } = this.props;
    fetchUserProfile(userId);
    fetchBookmarkedArticles();
  }

  render() {
    const { profileStore, bookmarked } = this.props;
    const fullName = `${profileStore.user.firstname} ${profileStore.user.lastname}`;
    if (profileStore.loading && !bookmarked) return <Loading />;
    return (
      <div>
        <div className="profile-details">
          <img alt="profile" className="profile-image" src={profileStore.profile.image || userPlaceholderImage} />
          <h3 id="user-name" className="username">{fullName}</h3>
        </div>
        <h3 className="profile-title">Bookmarks</h3>
        <BookmarkedArticleList articles={bookmarked} />
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  profileStore: state.profile,
  user: state.auth.ownProfile,
  bookmarked: state.articles.bookmarkedArticles,
});

export default connect(mapStateToProps, { fetchProfile, getBookmarkedArticles })(BookmarkedArticles);
