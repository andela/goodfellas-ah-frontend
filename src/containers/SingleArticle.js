import React, { Component } from 'react';
import { connect } from 'react-redux';
import { getAnArticle } from '../actions/articleActions';
import postComments from '../actions/postComments';
import getComments from '../actions/getComments';
import replyComments from '../actions/replyComments';
import ArticleHeader from '../components/article/ArticleHeader';
import ArticleBody from '../components/article/ArticleBody';
import Comment from '../components/article/Comment';


export class Article extends Component {
  state = {
    commentBody: '',
    replyBody: '',
  }

  componentWillMount = () => {
    const { match, getAnArticle: getArticle, getComments: getArticleComments } = this.props;
    const { slug } = match.params;

    if (slug) {
      getArticle(slug);
      getArticleComments(slug);
    }
  }

  handleChange = (event) => {
    const textarea = event.target;
    textarea.style.height = 'auto';
    textarea.style.height = `${textarea.scrollHeight}px`;
    const { name, value } = event.target;
    this.setState({ [name]: value });
  }

  handleKeyPress = (event) => {
    if (event.charCode === 13 && !event.shiftKey) {
      event.preventDefault();
      if (event.target.getAttribute('name') === 'replyBody') {
        const commentId = event.target.getAttribute('data-id');
        this.handleSubmitReply(commentId);
      } else {
        this.handleSubmitComment();
      }
    }
  }

  handleSubmitComment = async () => {
    const { commentBody: body } = this.state;
    const commentsPayload = {
      body,
    };
    const { match } = this.props;
    // eslint-disable-next-line react/destructuring-assignment
    this.props.postComments(commentsPayload, match.params.slug);
  }

  handleSubmitReply = async (commentId) => {
    const { replyBody: body } = this.state;
    const commentsPayload = {
      body,
    };
    // eslint-disable-next-line react/destructuring-assignment
    this.props.replyComments(commentsPayload, commentId);
  }

  singleArticle() {
    const { article, comments, userId } = this.props;
    return (
      <div>
        <div className="single-page">
          <ArticleHeader article={article} userId={userId} />
          <ArticleBody article={article} userId={userId} />
          <Comment
            article={article}
            comments={comments}
            handleKeyPress={this.handleKeyPress}
            handleChange={this.handleChange}
          />
        </div>
      </div>

    );
  }

  renderArticle() {
    const { article, error } = this.props;
    if (article) {
      return this.singleArticle();
    }
    return <div>{error}</div>;
  }


  render() {
    return (
      <div>
        { this.renderArticle() }
      </div>
    );
  }
}


const mapStateToProps = (state) => ({
  error: state.articles.articleError,
  article: state.articles.article,
  comment: state.postComments,
  replyComment: state.replyComments,
  comments: state.getComments.comments,
  userId: state.auth.ownProfile.userId,
});

export default connect(mapStateToProps, {
  replyComments, postComments, getComments, getAnArticle,
})(Article);
