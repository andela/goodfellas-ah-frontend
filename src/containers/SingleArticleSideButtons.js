import React, { Component, Fragment } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import icons from '../assets/icons.svg';
import { bookmarkArticles, removeBookmark } from '../actions/articleActions';

export class SideButtons extends Component {
  state = {
    extraButtons: false,
    showSideButton: true,
    bookmarked: false,
  }

  componentWillMount() {
    const { article } = this.props;
    if (article.bookmarked.length >= 1) {
      this.setAsBookMarked();
    }
  }

  componentDidUpdate(prevProps) {
    const { article } = this.props;
    if (article !== prevProps.article) {
      if (article.bookmarked.length >= 1) {
        this.setAsBookMarked();
      }
    }
  }

  setAsBookMarked() {
    this.setState({ bookmarked: true });
  }

  bookmarkAnArticle(bookmarked) {
    const {
      article: { slug },
      bookmarkArticles: bookmark,
      removeBookmark: removeArticleBookmark,
    } = this.props;

    if (bookmarked) {
      this.setState({ bookmarked: false });
      removeArticleBookmark(slug, (response) => {
        if (!response) { this.setState({ bookmarked: true }); }
      });
    } else {
      bookmark(slug, (response) => {
        this.setState({ bookmarked: true });
        if (!response) { this.setState({ bookmarked: false }); }
      });
    }
  }

  authorizedButtons = () => {
    const { userId, article } = this.props;
    if (userId === article.authorId) {
      return (
        <div>
          <svg id="delete" className="icon" title="Delete this article"><use xlinkHref={`${icons}#delete-icon`} /></svg>
          <svg id="edit" title="Edit this article"><Link className="link" to={`/articles/edit/${article.slug}`}><use xlinkHref={`${icons}#edit`} /></Link></svg>
        </div>
      );
    }
  }

  showExtraButtons = () => {
    this.setState((prevState) => ({ extraButtons: !prevState.extraButtons }));
  }

  toggleSideButton = () => {
    this.setState((prevState) => ({ showSideButton: !prevState.showSideButton }));
  }

  AddExtraButtons = (className) => (
    <div className={`extra-buttons ${className}`}>
      {this.authorizedButtons()}
      <svg id="flag" title="Report this article"><use xlinkHref={`${icons}#flag`} /></svg>
      <svg id="dislike" title="Dislike this article"><use xlinkHref={`${icons}#dislike`} /></svg>
    </div>
  )


  render() {
    const { extraButtons, showSideButton, bookmarked } = this.state;
    return (
      <Fragment>
        <span className="small-button" onClick={this.toggleSideButton}>click</span>
        <div className={`side-button ${showSideButton ? 'active' : ''}`}>
          <svg id="star" title="Like this article"><use xlinkHref={`${icons}#star`} /></svg>
          <svg id="bookmark" className={`not-bookmarked ${bookmarked ? 'bookmarked' : null}`} title="Report this article" onClick={() => this.bookmarkAnArticle(bookmarked)}><use xlinkHref={`${icons}#bookmark`} /></svg>
          <svg id="share" title="Share this article"><use xlinkHref={`${icons}#share`} /></svg>
          <svg id="more" onClick={this.showExtraButtons} title="Show more buttons"><use xlinkHref={`${icons}#more`} /></svg>
          {this.AddExtraButtons(extraButtons ? 'active' : '')}
          <div id="like-count"><p>4</p></div>
          <svg id="love" title="Report this article"><use xlinkHref={`${icons}#gratipay`} /></svg>
        </div>
      </Fragment>
    );
  }
}

const mapStateToProps = (state) => ({
  article: state.articles.article,
  userId: state.auth.ownProfile.userId,
});

export default connect(mapStateToProps, { bookmarkArticles, removeBookmark })(SideButtons);
