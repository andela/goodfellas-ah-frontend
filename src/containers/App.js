import React from 'react';
import { BrowserRouter, Route, Switch } from 'react-router-dom';
import '../styles/styles.scss';
import authenticate from './hoc/authenticate';
import LandingPage from '../views/LandingPage';
import Signin from '../views/SigninPage';
import Signup from '../views/SignupPage';
import SingleArticle from '../views/SingleArticlePage';
import EditProfile from '../views/ProfileEditPage';
import Profile from '../views/ProfilePage';
import SocialAuthPage from '../views/SocialAuthPage';
import ForgotPasswordPage from '../views/ForgotPasswordPage';
import ResetPasswordPage from '../views/ResetPasswordPage';
import SearchArticles from '../views/SearchArticles';
import NotificationPage from '../views/NotificationPage';
import GetAllArticles from '../views/AllArticles';
import Tags from '../components/article/Tags';
import CreateArticle from '../views/CreateArticles';
import UpdateArticle from '../views/UpdateArticle';
import BookmarkedArticles from '../views/BookmarkedArticlesPage';

const User = () => (
  <Switch>
    <Route exact path="/user/profile" component={authenticate(Profile)} />
    <Route exact path="/user/profile/edit" component={authenticate(EditProfile)} />
    <Route exact path="/user/profile/:userId" component={Profile} />
    <Route path="/user/notifications" component={NotificationPage} />
  </Switch>
);

const Auth = () => (
  <div>
    <Route path="/auth/signin" render={(props) => <Signin formtype="signin" {...props} />} />
    <Route path="/auth/signup" render={(props) => <Signup formtype="signup" {...props} />} />
    <Route path="/auth/social" component={SocialAuthPage} />
  </div>
);

const Articles = () => (
  <div>
    <Switch>
      <Route path="/articles/tags/create" component={Tags} />
      <Route path="/articles/search" component={SearchArticles} />
      <Route path="/articles/create" component={authenticate(CreateArticle)} />
      <Route path="/articles/bookmarked" component={authenticate(BookmarkedArticles)} />
      <Route path="/articles/home" component={GetAllArticles} />
      <Route path="/articles/edit/:slug" component={authenticate(UpdateArticle)} />
      <Route path="/articles/:slug" component={SingleArticle} />
    </Switch>
  </div>
);

const App = () => (
  <BrowserRouter>
    <div>
      <Route exact path="/" component={LandingPage} />
      <Route path="/auth" component={Auth} />
      <Route path="/articles" component={Articles} />
      <Route path="/user" component={User} />
      <Route path="/forgotpassword" component={ForgotPasswordPage} />
      <Route path="/resetpassword" component={ResetPasswordPage} />
    </div>
  </BrowserRouter>
);

export default App;
