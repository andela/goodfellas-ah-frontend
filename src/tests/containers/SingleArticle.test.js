import React from 'react';
import { shallow } from 'enzyme';
import { Article } from '../../containers/SingleArticle';

describe('containers', () => {
  describe('SingleArticle container', () => {
    const setup = () => {
      const props = {
        userId: '',
        article: {},
        error: '',
        getAnArticle: jest.fn(),
        getComments: jest.fn(),
        match: {
          params: { slug: 'some-slug' },
        },
      };

      const enzymeWrapper = shallow(<Article {...props} />);

      return {
        props,
        enzymeWrapper,
      };
    };
    it('should render as expected', () => {
      const { enzymeWrapper } = setup();
      expect(enzymeWrapper).toMatchSnapshot();
    });
  });
  describe('SingleArticle container :Error', () => {
    const setup = () => {
      const props = {
        userId: null,
        article: null,
        error: 'this is an error',
        getAnArticle: jest.fn(),
        getComments: jest.fn(),
        match: {
          params: { slug: 'some-slug' },
        },
      };

      const enzymeWrapper = shallow(<Article {...props} />);

      return {
        props,
        enzymeWrapper,
      };
    };
    it('should render as expected', () => {
      const { enzymeWrapper } = setup();
      expect(enzymeWrapper).toMatchSnapshot();
    });
  });
});
