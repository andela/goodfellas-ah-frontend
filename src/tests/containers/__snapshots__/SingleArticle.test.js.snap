// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers SingleArticle container :Error should render as expected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Article
    article={null}
    error="this is an error"
    getAnArticle={
      [MockFunction] {
        "calls": Array [
          Array [
            "some-slug",
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    getComments={
      [MockFunction] {
        "calls": Array [
          Array [
            "some-slug",
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    match={
      Object {
        "params": Object {
          "slug": "some-slug",
        },
      }
    }
    userId={null}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div>
        this is an error
      </div>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "this is an error",
      },
      "ref": null,
      "rendered": "this is an error",
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div>
          this is an error
        </div>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "this is an error",
        },
        "ref": null,
        "rendered": "this is an error",
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`containers SingleArticle container should render as expected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Article
    article={Object {}}
    error=""
    getAnArticle={
      [MockFunction] {
        "calls": Array [
          Array [
            "some-slug",
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    getComments={
      [MockFunction] {
        "calls": Array [
          Array [
            "some-slug",
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    match={
      Object {
        "params": Object {
          "slug": "some-slug",
        },
      }
    }
    userId=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div>
        <div
          className="single-page"
        >
          <ArticleHeader
            article={Object {}}
            userId=""
          />
          <ArticleBody
            article={Object {}}
            userId=""
          />
          <Comment
            article={Object {}}
            handleChange={[Function]}
            handleKeyPress={[Function]}
          />
        </div>
      </div>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="single-page"
        >
          <ArticleHeader
            article={Object {}}
            userId=""
          />
          <ArticleBody
            article={Object {}}
            userId=""
          />
          <Comment
            article={Object {}}
            handleChange={[Function]}
            handleKeyPress={[Function]}
          />
        </div>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <ArticleHeader
              article={Object {}}
              userId=""
            />,
            <ArticleBody
              article={Object {}}
              userId=""
            />,
            <Comment
              article={Object {}}
              handleChange={[Function]}
              handleKeyPress={[Function]}
            />,
          ],
          "className": "single-page",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "article": Object {},
              "userId": "",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "article": Object {},
              "userId": "",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "article": Object {},
              "comments": undefined,
              "handleChange": [Function],
              "handleKeyPress": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div>
          <div
            className="single-page"
          >
            <ArticleHeader
              article={Object {}}
              userId=""
            />
            <ArticleBody
              article={Object {}}
              userId=""
            />
            <Comment
              article={Object {}}
              handleChange={[Function]}
              handleKeyPress={[Function]}
            />
          </div>
        </div>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="single-page"
          >
            <ArticleHeader
              article={Object {}}
              userId=""
            />
            <ArticleBody
              article={Object {}}
              userId=""
            />
            <Comment
              article={Object {}}
              handleChange={[Function]}
              handleKeyPress={[Function]}
            />
          </div>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <ArticleHeader
                article={Object {}}
                userId=""
              />,
              <ArticleBody
                article={Object {}}
                userId=""
              />,
              <Comment
                article={Object {}}
                handleChange={[Function]}
                handleKeyPress={[Function]}
              />,
            ],
            "className": "single-page",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "article": Object {},
                "userId": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "article": Object {},
                "userId": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "article": Object {},
                "comments": undefined,
                "handleChange": [Function],
                "handleKeyPress": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
